app/src/main/java/net/bonysoft/magicmirror/FaceRecognitionActivity.java

```diff
@@ -50,22 +54,35 @@ protected void onCreate(Bundle savedInstanceState) {
             return;
         }
 
-        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
-            ActivityCompat.requestPermissions(
-                    this,
-                    new String[]{Manifest.permission.CAMERA},
-                    CAMERA_PERMISSION_REQUEST
-            );
+        if (isUsingCamera()) {
+            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
+                ActivityCompat.requestPermissions(
+                        this,
+                        new String[]{Manifest.permission.CAMERA},
+                        CAMERA_PERMISSION_REQUEST
+                );
+            } else {
+                tryToCreateCameraSource();
+            }
+            displayErrorIfPlayServicesMissing();
         } else {
-            tryToCreateCameraSource();
+            createKeyboardSource();
         }
-        displayErrorIfPlayServicesMissing();
+    }
+
+    private boolean isUsingCamera() {
```

Comment:
`isPhysicalDevice`, given that you can use the camera in an emulator.
Here we want to check if the device is an emulator or not